clang/lib/Sema/SemaType.cpp

```diff
@@ -8240,59 +8240,56 @@ static void HandleCHERICapabilityQualifier(QualType &CurType,
       // counterparts; all other cases are shared with the declarator chunk
       // case at the end of the function.
       if (const TypedefType *TT = CurType->getAs<TypedefType>()) {
-        CurType = S.Context.getTypedefType(TT->getDecl(), QualType(), true);
-        return;
+        break;
       }
-    } else {
-      for (unsigned i = state.getCurrentChunkIndex(); i != 0; --i) {
-        DeclaratorChunk &chunk = declarator.getTypeObject(i-1);
-        switch (chunk.Kind) {
-          case DeclaratorChunk::Pointer: {
-            // Check for an ambiguous use where this is more than one pointer
-            // level, like __capability T **. We do this by checking that the next
-            // chunk isn't a pointer without the attribute.
-            if (i > 1) {
-              DeclaratorChunk &nextChunk = declarator.getTypeObject(i-2);
-              if (nextChunk.Kind == DeclaratorChunk::Pointer) {
-                auto Attr = nextChunk.getAttrs();
-                if (!Attr.hasAttribute(ParsedAttr::AT_CHERICapability)) {
-                  S.Diag(nextChunk.Loc,
-                         diag::err_cheri_capability_qualifier_ambiguous);
-                  return;
-                }
-              }
+    }
+    for (unsigned i = state.getCurrentChunkIndex(); i != 0; --i) {
```

Comment:
Hm, what are the consequences of not having this in the else? (Is it even possible to have a pointer or reference here that's not a typedef?)

Reply:
I'm not sure if this can happen. If you prefer I can restore the existing logic that does not enter the loop for pointer/reference types.

Reply:
I think so, and you can kill the typedef check then too as the control flow becomes equivalent?

Reply:
Looks like it can also happen with typeof(), I'll extend the test coverage.